### PR TITLE
fix: emit redactor-proof structured fields on telemetry errors

### PR DESCRIPTION
**Pull request by ralphstodomingo**
## Summary

VS Code's `TelemetryLogger` PII redactor whole-replaces any string property that matches one of its hardcoded patterns (URL, Generic Secret, email, JWT, GitHub token, user-file-path, etc.). When that fires on `stack` or `error_message` the event becomes `<REDACTED: ...>` — clusterless, unactionable. The pattern set lives in VS Code core and grows over time.

Chasing it with our own keyword allowlist was an arms race we cannot win — incomplete by construction (~6% of staging error events still got whole-redacted under the previous Bearer / JWT / keyword=value sanitizer), and the list of keywords would have grown forever as new credential formats appeared.

Replace that pattern-chase with a fixed set of **atomic structured fields the redactor cannot match by construction**, added to every `sendTelemetryError` payload:

| Field | Value | Why redactor-safe |
|---|---|---|
| `stack_hash` | sha1 of normalised frames (`fn\|basename\|line`) | 40 hex chars — matches no pattern |
| `stack_top_frame_fn` | function/method name of first non-internal frame | enum-shape token, no `://`, no `@`, no secret keyword |
| `stack_top_frame_file` | basename only (e.g. `extension.js`) | no path, no scheme |
| `error_constructor` | `error.constructor.name` | `Error`, `TypeError`, `PythonException`, ... |
| `is_power_user_origin` | `"true"` / `"false"` | enum token |
| `stack_top_frame_line` *(measurement)* | line number | numeric, never goes through string redactor |
| `stack_frame_count` *(measurement)* | total parsed frames | numeric |

App Insights grouping / alerting moves to `stack_hash` + `stack_top_frame_fn`. The readable `stack` and `error_message` stay as best-effort strings — if VS Code masks them in one event we still cluster and triage from the structured fields.

Also trims `sanitizeForTelemetry` to a small **URL + email safety net** (URL regex broadened to VS Code's `[a-zA-Z][a-zA-Z0-9+.-]*://` so any scheme — `file://`, `vscode://`, `postgres://` — is pre-masked, mirroring VS Code's span). Drops Bearer / JWT / keyword=value masking entirely; the structured fields make those unnecessary.

## Why this scope, with evidence

48h on staging (appId `429da6f5…`):

| Slice | Events | Notes |
|---|---:|---|
| `catchAllError` total | 33,984 | |
| any redaction marker in stack | 33,961 (99.93%) | almost everything has some redaction |
| **whole-field redacted stack** | 1,963 (5.8%) | the residual a sanitizer-only fix cannot close |
| inline-redacted only (paths masked, frames readable) | 31,998 (94.2%) | fn names + line numbers + post-`.vscode/` path survive |

The redaction problem is broader than `catchAllError`. The same `sendTelemetryError` path runs through every error event, and the same residual surfaces on each:

| Event | Redacted events / machines |
|---|---|
| catchAllError | 32,155 / 1,547 |
| DBTProjectIntegrationAdapter | 4,541 / 543 |
| Lineage:getMissingLineageMessage | 1,936 / 175 |
| dbtCore:executeSQL | 463 / 162 |
| pythonBridgeInitPythonError | 325 / 83 |
| ManifestParser | 631 / 28 |
| formatDbtModelApplyDiffError | 204 / 29 |

Structured fields fix the residual on **all** of these, not just `catchAllError`.

## Triage post-merge — query shape

App Insights clustering moves from substring-matching a possibly-masked stack to:

```kql
customEvents
| where timestamp > ago(24h) and name endswith "Error"
| extend stack_hash         = tostring(customDimensions.stack_hash),
         top_frame          = tostring(customDimensions.stack_top_frame_fn),
         top_frame_file     = tostring(customDimensions.stack_top_frame_file),
         error_ctor         = tostring(customDimensions.error_constructor),
         pu_origin          = tostring(customDimensions.is_power_user_origin)
| where pu_origin == "true"
| summarize machines = dcount(tostring(customDimensions.['common.vscodemachineid'])),
            events   = count()
            by stack_hash, top_frame, top_frame_file, error_ctor
| order by machines desc
```

Every cluster has a top-frame name and basename + line + constructor — enough to find the code without the readable stack.

## Test plan

- [x] 28 unit tests, covering:
  - Existing `error_name` / `error_message` / `error_code` extraction (no regression)
  - New `error_constructor` for `Error`, `TypeError`, `RangeError`, non-Error values, and `undefined`
  - `is_power_user_origin` true/false from extension-dir marker in stack
  - Top-frame extraction skips `node:internal/...` and `<anonymous>`, prefers `extension.js` frames
  - `stack_frame_count` is a **measurement** (numeric), not a string property — confirms it doesn't go through the string redactor
  - `stack_hash` stability across Windows vs Unix user paths for the same logical stack
  - `stack_hash` differs for different logical stacks
  - Survives stacks that VS Code's `user-file-path` redactor already touched inline
  - No-stack / non-Error / undefined / null fallbacks
  - Slim `sanitizeForTelemetry`: matches any-scheme URLs (`file://`, `postgres://`, `vscode-resource://`), masks emails, does NOT mask `node:internal/*` (single slash, not `://`), no Bearer / JWT / keyword=value masking
- [x] Full jest suite green (509 passed / 1 skipped, no regressions)
- [x] ESLint + `tsc --noEmit` clean
- [x] Bundle builds via `rsbuild build --mode development` (5.4 MB extension.js, contains `stack_hash` / `stack_top_frame_fn` / `is_power_user_origin` / `error_constructor`; old `JWT_RE` / `BEARER_RE` / `SECRET_KV_RE` identifiers no longer present)

## Relationship to #1949

Independent and complementary. #1949 filters out rejections from other extensions and python-bridge cleanup races at the unhandledRejection handler. This PR fixes the diagnosability of every error event that does flow through, regardless of which call site emitted it.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Chores**
  * Enhanced error telemetry: safer sanitization of messages/stack traces, stable anonymized stack identifiers, richer structured metadata, and more robust disposal handling
  * Refined diagnostics pipeline for more consistent and redaction-resilient telemetry

* **Tests**
  * Expanded test coverage for telemetry parsing, sanitization, anonymization, and numerous edge cases (non-Error inputs, circular values, redacted stacks)

### CODE REVIEW

File: src/telemetry/index.ts [hunk 1]
```diff
@@ -135,19 +143,160 @@ export class TelemetryService implements vscode.Disposable {
...
+    return filePath;
+  }
+  // VS Code's user-file-path redactor may have already replaced the
+  // absolute path with "<REDACTED: user-file-path>". The remaining
+  // ".../foo/bar.js" tail still has a real basename worth keeping.
+  const m = /([^\\/]+?)$/.exec(filePath);
+  return m ? m[1] : filePath;
 }
+
+function pickTopFrame(frames: StackFrame[]): StackFrame | undefined {
```
Comment (saravmajestic) [inline src/telemetry/index.ts:268]: Nit: Can we match some marker text in the JS file to be sure that it is from PU extension?

**Review by saravmajestic — APPROVED** [review]

File: src/telemetry/index.ts (hunk 1 above)
Comment (ralphstodomingo) [reply src/telemetry/index.ts:268]: Done in 6544841a — frames now keep their full path and `pickTopFrame` matches the `innoverio.vscode-dbt-power-user` marker instead of the `extension.js` basename, so the top frame is only attributed to us when the path is actually from our extension dir (the marker survives even when VS Code's redactor masks the home-dir prefix). Added a mixed-extension-stack test (gitlens frame first) to lock it in.

File: src/telemetry/index.ts [hunk 2]
```diff
@@ -135,19 +143,160 @@ export class TelemetryService implements vscode.Disposable {
...
+  const powerUser = frames.find(
+    (f) => !f.isInternal && f.file.includes("extension.js"),
+  );
+  if (powerUser) {
+    return powerUser;
+  }
+  return frames.find((f) => !f.isInternal) || frames[0];
```
[the remark this answers is not in the document]
Comment (ralphstodomingo) [reply src/telemetry/index.ts:278]: Fixed in 6544841a — `StackFrame` now carries `fullPath` and `pickTopFrame` matches `POWER_USER_EXTENSION_MARKER` on it, with the basename only used for reporting. Covered by a new mixed-extension-stack test plus a no-marker fallback test.

File: src/telemetry/index.ts [hunk 3]
```diff
@@ -88,6 +91,11 @@ export class TelemetryService implements vscode.Disposable {
     properties?: { [key: string]: string },
     measurements?: { [key: string]: number },
   ) {
+    const rawStack =
+      error !== undefined && error instanceof Error
+        ? error.stack
+        : JSON.stringify(error);
+    const structured = extractStructuredFields(error, rawStack);
```
[the remark this answers is not in the document]
Comment (ralphstodomingo) [reply src/telemetry/index.ts:98]: Fixed in 6544841a — non-Error throwables are serialized via a guarded `safeSerializeTelemetryValue` (JSON.stringify with `String(value)` fallback), so circular structures and BigInt no longer throw inside `sendTelemetryError`. The top-frame attribution half of this comment is addressed together with the line-271 thread (marker match on the full path). Tests added for both circular and BigInt values.
